File: crates/phi-daemon/src/runtime/ask_user.rs
use std::collections::HashSet;

use async_trait::async_trait;
use phi::{Tool, ToolDefinition, ToolEffect, ToolError, ToolOutput};
use serde::{Deserialize, Serialize};
use serde_json::json;
use tokio::sync::{mpsc, oneshot};

use super::AskUserId;

pub const ASK_USER_TOOL_NAME: &str = "askuser";

const ASK_USER_DESCRIPTION: &str = r#"Use this tool only when you are blocked on a decision that is genuinely the user's to make: one you cannot resolve from the request, the code, or sensible defaults.

Usage notes:
- Users will always be able to select "Other" to provide custom text input
- Use multiSelect: true to allow multiple answers to be selected for a question
- If you recommend a specific option, make that the first option in the list and add "(Recommended)" at the end of the label
- Do not use this tool merely to ask whether to proceed or whether a plan is ready

Reserve this for decisions where the user's answer changes what you do next — not for choices with a conventional default or facts you can verify in the codebase yourself. In those cases pick the obvious option, mention it in your response, and proceed.

Preview feature:
Use the optional preview field on options when presenting concrete artifacts that users need to visually compare:
- ASCII mockups of UI layouts or components
- Code snippets showing different implementations
- Diagram variations
- Configuration examples

Preview content is rendered as markdown in a monospace box. Multi-line text with newlines is supported. When any option has a preview, the UI switches to a side-by-side layout with a vertical option list on the left and preview on the right. Do not use previews for simple preference questions where labels and descriptions suffice. Previews are only supported for single-select questions."#;

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AskUserOption {
    pub label: String,
    pub description: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub preview: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AskUserQuestion {
    pub question: String,
    pub header: String,
    pub options: Vec<AskUserOption>,
    #[serde(default)]
    pub multi_select: bool,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct AskUserRequest {
    pub ask_id: AskUserId,
    pub questions: Vec<AskUserQuestion>,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub struct AskUserAnswer {
    pub question_index: usize,
    #[serde(default)]
    pub selected_options: Vec<String>,
    #[serde(default)]
    pub custom_text: Option<String>,
}

#[derive(Debug, Deserialize)]
struct AskUserArguments {
    questions: Vec<AskUserQuestion>,
}

pub(super) struct PendingAskUserRequest {
    pub request: AskUserRequest,
    pub reply: oneshot::Sender<Result<Vec<AskUserAnswer>, String>>,
}

#[derive(Clone)]
pub(super) struct AskUserTool {
    requests: mpsc::UnboundedSender<PendingAskUserRequest>,
}

impl AskUserTool {
    pub(super) fn channel() -> (Self, mpsc::UnboundedReceiver<PendingAskUserRequest>) {
        let (requests, receiver) = mpsc::unbounded_channel();
        (Self { requests }, receiver)
    }
}

#[async_trait]
impl Tool for AskUserTool {
    fn definition(&self) -> ToolDefinition {
        ToolDefinition::new(
            ASK_USER_TOOL_NAME,
            ASK_USER_DESCRIPTION,
            json!({
                "type": "object",
                "additionalProperties": false,
                "properties": {
                    "questions": {
                        "type": "array",
                        "minItems": 1,
                        "maxItems": 3,
                        "items": {
                            "type": "object",
                            "additionalProperties": false,
                            "properties": {
                                "question": { "type": "string", "minLength": 1 },
                                "header": { "type": "string", "minLength": 1, "maxLength": 12 },
                                "options": {
                                    "type": "array",
                                    "minItems": 2,
                                    "maxItems": 4,
                                    "items": {
                                        "type": "object",
                                        "additionalProperties": false,
                                        "properties": {
                                            "label": { "type": "string", "minLength": 1 },
                                            "description": { "type": "string", "minLength": 1 },
                                            "preview": { "type": "string", "minLength": 1 }
                                        },
                                        "required": ["label", "description"]
                                    }
                                },
                                "multiSelect": { "type": "boolean", "default": false }
                            },
                            "required": ["question", "header", "options"]
                        }
                    }
                },
                "required": ["questions"]
            }),
        )
    }

    fn effect(&self) -> ToolEffect {
        ToolEffect::Internal
    }

    async fn execute(&self, arguments: serde_json::Value) -> Result<ToolOutput, ToolError> {
        let arguments: AskUserArguments = serde_json::from_value(arguments)
            .map_err(|error| ToolError::new(format!("invalid askuser arguments: {error}")))?;
        validate_questions(&arguments.questions).map_err(ToolError::new)?;

        let request = AskUserRequest {
            ask_id: AskUserId::new(),
            questions: arguments.questions,
        };
        let (reply, response) = oneshot::channel();
        self.requests
            .send(PendingAskUserRequest {
                request: request.clone(),
                reply,
            })
            .map_err(|_| ToolError::new("askuser runtime is unavailable"))?;
        let answers = response
            .await
            .map_err(|_| ToolError::new("askuser request was cancelled"))?
            .map_err(ToolError::new)?;

        Ok(ToolOutput::success(format_answers(
            &request.questions,
            &answers,
        )?))
    }
}

fn validate_questions(questions: &[AskUserQuestion]) -> Result<(), String> {
    if !(1..=3).contains(&questions.len()) {
        return Err("askuser requires between 1 and 3 questions".to_owned());
    }
    for (index, question) in questions.iter().enumerate() {
        if question.question.trim().is_empty() {
            return Err(format!("question {index} must not be empty"));
        }
        let header_length = question.header.chars().count();
        if !(1..=12).contains(&header_length) {
            return Err(format!(
                "question {index} header must contain between 1 and 12 characters"
            ));
        }
        if !(2..=4).contains(&question.options.len()) {
            return Err(format!("question {index} requires between 2 and 4 options"));
        }
        let mut labels = HashSet::with_capacity(question.options.len());
        for option in &question.options {
            if option.label.trim().is_empty() {
                return Err(format!("question {index} has an empty option label"));
            }
            if option.description.trim().is_empty() {
                return Err(format!(
                    "question {index} option {:?} has an empty description",
                    option.label
                ));
            }
            if !labels.insert(option.label.as_str()) {
                return Err(format!(
                    "question {index} has duplicate option label {:?}",
                    option.label
                ));
            }
            if option
                .preview
                .as_ref()
                .is_some_and(|preview| preview.is_empty())
            {
                return Err(format!(
                    "question {index} option {:?} has an empty preview",
                    option.label
                ));
            }
            if question.multi_select && option.preview.is_some() {
                return Err(format!(
                    "question {index} cannot use previews when multiSelect is true"
                ));
            }
        }
    }
    Ok(())
}

pub(super) fn validate_answers(
    questions: &[AskUserQuestion],
    answers: &[AskUserAnswer],
) -> Result<(), String> {
    if answers.len() != questions.len() {
        return Err(format!(
            "expected {} answers, received {}",
            questions.len(),
            answers.len()
        ));
    }
    let mut answered = HashSet::with_capacity(answers.len());
    for answer in answers {
        let Some(question) = questions.get(answer.question_index) else {
            return Err(format!(
                "question_index {} is out of range",
                answer.question_index
            ));
        };
        if !answered.insert(answer.question_index) {
            return Err(format!(
                "question_index {} was answered more than once",
                answer.question_index
            ));
        }
        let mut selected = HashSet::with_capacity(answer.selected_options.len());
        for label in &answer.selected_options {
            if !selected.insert(label.as_str()) {
                return Err(format!(
                    "question_index {} selected option {:?} more than once",
                    answer.question_index, label
                ));
            }
            if !question.options.iter().any(|option| option.label == *label) {
                return Err(format!(
                    "question_index {} selected unknown option {:?}",
                    answer.question_index, label
                ));
            }
        }
        let has_custom_text = match &answer.custom_text {
            Some(custom_text) if custom_text.trim().is_empty() => {
                return Err(format!(
                    "question_index {} custom_text must not be empty",
                    answer.question_index
                ));
            }
            Some(_) => true,
            None => false,
        };
        let selection_count = answer.selected_options.len() + usize::from(has_custom_text);
        if selection_count == 0 {
            return Err(format!(
                "question_index {} requires an option or custom_text",
                answer.question_index
            ));
        }
        if !question.multi_select && selection_count != 1 {
            return Err(format!(
                "question_index {} is single-select and requires exactly one answer",
                answer.question_index
            ));
        }
    }
    Ok(())
}

fn format_answers(
    questions: &[AskUserQuestion],
    answers: &[AskUserAnswer],
) -> Result<String, ToolError> {
    #[derive(Serialize)]
    struct ResultAnswer<'a> {
        question: &'a str,
        selected_options: &'a [String],
        custom_text: Option<&'a str>,
    }

    #[derive(Serialize)]
    struct ResultEnvelope<'a> {
        answers: Vec<ResultAnswer<'a>>,
    }

    let answers = questions
        .iter()
        .enumerate()
        .map(|(index, question)| {
            let answer = answers
                .iter()
                .find(|answer| answer.question_index == index)
                .expect("validated askuser answers must cover every question");
            ResultAnswer {
                question: &question.question,
                selected_options: &answer.selected_options,
                custom_text: answer.custom_text.as_deref(),
            }
        })
        .collect();
    serde_json::to_string(&ResultEnvelope { answers })
        .map_err(|error| ToolError::new(format!("could not serialize askuser answers: {error}")))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn question(multi_select: bool) -> AskUserQuestion {
        AskUserQuestion {
            question: "Which approach?".to_owned(),
            header: "Approach".to_owned(),
            options: vec![
                AskUserOption {
                    label: "A".to_owned(),
                    description: "Use A".to_owned(),
                    preview: None,
                },
                AskUserOption {
                    label: "B".to_owned(),
                    description: "Use B".to_owned(),
                    preview: None,
                },
            ],
            multi_select,
        }
    }

    #[test]
    fn accepts_a_custom_other_answer() {
        let questions = [question(false)];
        let answers = [AskUserAnswer {
            question_index: 0,
            selected_options: Vec::new(),
            custom_text: Some("My own choice".to_owned()),
        }];

        validate_answers(&questions, &answers).unwrap();
    }

    #[test]
    fn rejects_multiple_values_for_a_single_select_question() {
        let questions = [question(false)];
        let answers = [AskUserAnswer {
            question_index: 0,
            selected_options: vec!["A".to_owned(), "B".to_owned()],
            custom_text: None,
        }];

        assert!(validate_answers(&questions, &answers).is_err());
    }
}

File: pi-ext/src/ask-user.ts
import { defineTool, type ToolDefinition } from "@earendil-works/pi-coding-agent";
import { uuidv7 } from "@earendil-works/pi-ai";
import { Type } from "typebox";

import { CommandError } from "./errors.js";
import type { AskUserAnswer, AskUserQuestion, AskUserRequest } from "./protocol.js";

const AskUserParameters = Type.Object({
  questions: Type.Array(
    Type.Object({
      question: Type.String({ minLength: 1 }),
      header: Type.String({ minLength: 1, maxLength: 12 }),
      options: Type.Array(
        Type.Object({
          label: Type.String({ minLength: 1 }),
          description: Type.String({ minLength: 1 }),
          preview: Type.Optional(Type.String({ minLength: 1 })),
        }),
        { minItems: 2, maxItems: 4 },
      ),
      multiSelect: Type.Optional(Type.Boolean({ default: false })),
    }),
    { minItems: 1, maxItems: 3 },
  ),
});
const MAX_PENDING_ASKS = 64;

interface PendingAsk {
  request: AskUserRequest;
  resolve: (answer: string) => void;
  reject: (error: Error) => void;
  removeAbortListener: () => void;
}

export interface AskUserBrokerEvents {
  requested(request: AskUserRequest): void;
  answered(askId: string): void;
  cancelled(askId: string): void;
}

export class AskUserBroker {
  readonly #pending = new Map<string, PendingAsk>();
  #events: AskUserBrokerEvents | undefined;

  bind(events: AskUserBrokerEvents): void {
    this.#events = events;
  }

  createTool(): ToolDefinition {
    return defineTool({
      name: "askuser",
      label: "Ask user",
      description:
        "Ask the user for one to three decisions that are genuinely theirs to make. The request remains pending across client reconnects.",
      parameters: AskUserParameters,
      execute: async (_toolCallId, parameters, signal) => {
        const questions = normalizeQuestions(parameters.questions);
        const answer = await this.#request(questions, signal);
        return {
          content: [{ type: "text", text: answer }],
          details: {},
        };
      },
    });
  }

  listPending(): AskUserRequest[] {
    return [...this.#pending.values()].map(({ request }) => structuredClone(request));
  }

  answer(askId: string, answers: AskUserAnswer[]): void {
    const pending = this.#pending.get(askId);
    if (pending === undefined) {
      throw new CommandError("askuser_not_pending", `askuser request \`${askId}\` is not pending`);
    }
    const normalized = validateAnswers(pending.request.questions, answers);
    this.#pending.delete(askId);
    pending.removeAbortListener();
    pending.resolve(formatAnswers(pending.request.questions, normalized));
    this.#events?.answered(askId);
  }

  cancelAll(reason = "askuser request was cancelled"): void {
    for (const askId of [...this.#pending.keys()]) this.#cancel(askId, reason);
  }

  #request(questions: AskUserQuestion[], signal?: AbortSignal): Promise<string> {
    if (signal?.aborted) return Promise.reject(new Error("askuser request was cancelled"));
    if (this.#pending.size >= MAX_PENDING_ASKS) {
      return Promise.reject(new Error("too many askuser requests are pending"));
    }
    const askId = uuidv7();
    const request: AskUserRequest = { ask_id: askId, questions };
    return new Promise<string>((resolve, reject) => {
      const onAbort = (): void => this.#cancel(askId, "askuser request was cancelled");
      signal?.addEventListener("abort", onAbort, { once: true });
      this.#pending.set(askId, {
        request,
        resolve,
        reject,
        removeAbortListener: () => signal?.removeEventListener("abort", onAbort),
      });
      this.#events?.requested(structuredClone(request));
    });
  }

  #cancel(askId: string, reason: string): void {
    const pending = this.#pending.get(askId);
    if (pending === undefined) return;
    this.#pending.delete(askId);
    pending.removeAbortListener();
    pending.reject(new Error(reason));
    this.#events?.cancelled(askId);
  }
}

function normalizeQuestions(
  questions: Array<Omit<AskUserQuestion, "multiSelect"> & { multiSelect?: boolean }>,
): AskUserQuestion[] {
  if (questions.length < 1 || questions.length > 3) {
    throw new Error("askuser requires between 1 and 3 questions");
  }
  return questions.map((question, index) => {
    const normalized: AskUserQuestion = {
      question: question.question,
      header: question.header,
      options: question.options.map((option) => ({
        label: option.label,
        description: option.description,
        ...(option.preview === undefined ? {} : { preview: option.preview }),
      })),
      multiSelect: question.multiSelect ?? false,
    };
    if (!normalized.question.trim()) throw new Error(`question ${index} must not be empty`);
    const headerLength = Array.from(normalized.header).length;
    if (headerLength < 1 || headerLength > 12) {
      throw new Error(`question ${index} header must contain between 1 and 12 characters`);
    }
    if (normalized.options.length < 2 || normalized.options.length > 4) {
      throw new Error(`question ${index} requires between 2 and 4 options`);
    }
    const labels = new Set<string>();
    for (const option of normalized.options) {
      if (!option.label.trim() || !option.description.trim()) {
        throw new Error(`question ${index} option labels and descriptions must not be empty`);
      }
      if (option.preview === "") {
        throw new Error(`question ${index} option previews must not be empty`);
      }
      if (labels.has(option.label)) {
        throw new Error(`question ${index} has duplicate option label \`${option.label}\``);
      }
      labels.add(option.label);
      if (normalized.multiSelect && option.preview !== undefined) {
        throw new Error(`question ${index} cannot use previews when multiSelect is true`);
      }
    }
    return normalized;
  });
}

function validateAnswers(
  questions: readonly AskUserQuestion[],
  answers: readonly AskUserAnswer[],
): AskUserAnswer[] {
  if (answers.length !== questions.length) {
    throw new CommandError(
      "invalid_askuser_answer",
      `expected ${questions.length} answers, received ${answers.length}`,
    );
  }
  const byIndex = new Map<number, AskUserAnswer>();
  for (const answer of answers) {
    if (
      !Number.isSafeInteger(answer.question_index) ||
      answer.question_index < 0 ||
      answer.question_index >= questions.length ||
      byIndex.has(answer.question_index) ||
      !Array.isArray(answer.selected_options) ||
      (answer.custom_text !== null && typeof answer.custom_text !== "string")
    ) {
      throw new CommandError("invalid_askuser_answer", "askuser answer shape is invalid");
    }
    const question = questions[answer.question_index];
    if (question === undefined) {
      throw new CommandError("invalid_askuser_answer", "askuser question index is invalid");
    }
    const labels = new Set(question.options.map((option) => option.label));
    if (answer.selected_options.some((option) => !labels.has(option))) {
      throw new CommandError("invalid_askuser_answer", "askuser answer selected an unknown option");
    }
    if (new Set(answer.selected_options).size !== answer.selected_options.length) {
      throw new CommandError("invalid_askuser_answer", "askuser answer repeated an option");
    }
    if (answer.custom_text !== null && !answer.custom_text.trim()) {
      throw new CommandError("invalid_askuser_answer", "askuser custom answer must not be empty");
    }
    const answerCount = answer.selected_options.length + (answer.custom_text === null ? 0 : 1);
    if (answerCount === 0) {
      throw new CommandError("invalid_askuser_answer", "askuser answer must not be empty");
    }
    if (!question.multiSelect && answerCount !== 1) {
      throw new CommandError(
        "invalid_askuser_answer",
        "a single-select askuser question requires exactly one answer",
      );
    }
    byIndex.set(answer.question_index, {
      question_index: answer.question_index,
      selected_options: [...answer.selected_options],
      custom_text: answer.custom_text,
    });
  }
  return [...byIndex.values()].sort((left, right) => left.question_index - right.question_index);
}

function formatAnswers(
  questions: readonly AskUserQuestion[],
  answers: readonly AskUserAnswer[],
): string {
  const rendered = answers.map((answer) => ({
    question: questions[answer.question_index]?.question ?? "",
    selected_options: answer.selected_options,
    custom_text: answer.custom_text,
  }));
  return JSON.stringify({ answers: rendered });
}
